### ISSUE TITLE
gemma MoE prefill routes through cuBLAS matmul (m-dependent) — expert selection depends on the prime chunk/tick shape; 26B-A4B prefill logits move O(1) with MEMRA_PRIME_CHUNK, row 0

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#562]
## The defect

The gemma MoE arm routes prefill tokens through the batched cuBLAS matmul:

```rust
// crates/memra-engine/src/hybrid_forward.rs, gemma4_moe (tree 2394110f)
let logits = if t < PRIME_MIN_T {
    e.router_gemv(m.gate_inp.float_data(), router_in, n_embd, n_expert, t)?
} else {
    e.matmul(&m.gate_inp, router_in, t)?      // cuBLAS: reduction order depends on m
};
```

cuBLAS(Lt) selects its algorithm by shape, so the router logits of a token depend on **how many rows were in the call** — the prime chunk it happened to land in, or the tick budget. Top-k over 128 experts has near-ties; a last-ulp move flips the expert SET for that token, and 48 MoE layers amplify it to O(1) at the logits.

**Measured** (gemma-4-26B-A4B QAT Q4_0, T=4882, rented 5090, `research/exec-p1a-gemma-prime-20260919/raw/26b-a4b/`): with the generic chunked driver allowed to split, prefill logits differ at EVERY chunk size (1024/513/256/64 vs 4096; maxdiff 3.6–8.2), **first divergence at row 0** (rules out boundary carry — row 0 sees no cross-chunk state), greedy streams part at step 4–8. Multi-call `tickinv`: 0/7 EXACT. **Attribution** (`raw/diag-26b/`): unchanged with `MEMRA_GEMMA_MOE_MMA=0` (dp4a expert pairs instead of the int8-MMA expert GEMM) → the expert kernels are not the mover; the router is.

The dense gemma arms (12B, 31B) are bit-identical across the same splits, so this is the MoE arm alone.

## The serial trunk already closed this class

`moe_router_logits` (lane/concat-prime-exact): *"Real prefill must use the row-wise GEMV when exact routing is enabled: cuBLASLt's reduction changes with `m`, which makes expert selection depend on the caller's chunk or concat-batch shape."* `router_gemv`'s batch twin is bit-identical per row to the w8 form at every m (kernel-check sweeps m=1..2048), so one routing program can serve decode, verify and any prime split. The gemma arm never got that fix because its prime was never chunked before #561.

## Fix (in PR #561)

`gemma4_moe` routes through `router_gemv` at every `t`. Gate: 26B `chunk-invariance-gate.sh` (4096..64) and `tickinv` (5/20/77 calls + off-grid resumes) must read EXACT; argmax-margin re-run (the router's FP order changes — new routing program for gemma prefill, same as decode/verify already use). If EXACT, the registry row `GemmaParallelMoeResidual.chunked_prime` flips to yes and the 26B leaves the monolithic-prime class with the dense gemmas.

## Same class, other site (not fixed here)

`moe_ffn_lockstep` (`hybrid_forward.rs:~21813`, `let logits = e.matmul(&m.gate_inp, zbatch, mrows)?`) routes ALL lockstep-decode stream rows in one cuBLAS call with m = number of streams — a session's expert selection then depends on how many peers share its lockstep batch. Reachable from `run_lockstep` (CLI); not on the serving worker's path (`decode_step_lockstep` has no server caller). It should move to `router_gemv` for the same reason when that path is next touched.

Refs: #535 (P1a), #534, PR #561.

**Comment by avifenesh** [comment]
Root cause confirmed and fixed in PR #561 @ `ad1b0f58`.

**Attribution** (`research/exec-p1a-gemma-prime-20260919/raw/26b-a4b/diag/`): chunk dependence unchanged with `MEMRA_GEMMA_MOE_MMA=0` → expert kernels are not the mover; the cuBLAS router matmul is.

**Fix**: `gemma4_moe` routes through `router_gemv` at every `t` (mirrors the serial trunk's `moe_router_logits`).

**Receipt** (`raw/26b-a4b/routerfix/`, chunking allowed on the 26B): chunkinv **EXACT** at 4096/1024/513/256/64; tickinv **EXACT** at 5/20/77 calls + off-grid resumes 64/256/512/1000. Registry row `GemmaParallelMoeResidual.chunked_prime` → yes; the 26B leaves the monolithic-prime class. Dense 12B control unchanged; fixture gate 4/4; memra-gguf 280 / memra-server 711 tests pass.

argmax-margin on the fixed 26B: flips=2, both margin-explained (0.014 / 0.0135 vs spreads 0.64 / 3.02), decision position agreed. The 26B had no calibration row (default 1/12); a `gemma-4-26B*` row (3/12, from the measured decode margins — identical across both binaries, 5–6 of 12 positions arithmetically flippable, the 31B's distribution) is added in the same PR.

**Left open by this issue**: `moe_ffn_lockstep` (`hybrid_forward.rs`, `e.matmul(&m.gate_inp, zbatch, mrows)`) — lockstep decode routes all stream rows through one cuBLAS call with m = stream count, so a session's expert selection depends on its peer count. CLI `run_lockstep` only, no serving caller. Move to `router_gemv` when that path is next touched; keep this issue open for that.

**Comment by avifenesh** [comment]
Closing: both sites are on the m-invariant selector on main.

- gemma prefill (`gemma4_moe`): `router_gemv` at every `t` since #561 (`746bba2b4`), chunk-invariance and tick-invariance EXACT on the 26B.
- lockstep (`moe_ffn_lockstep`): `moe_router_logits` / `router_gemv` since #565 (`21c5f932c`), with the owed `run_lockstep` A/B on Hy3 (`research/lockstep-router-ab-20260920/`): main's M=4 same-prompt run diverged from M=1 at token 18, the fix reproduces M=1 on every stream. The two further peer-dependent sites that A/B surfaced (shared-expert `mrows` matmul, CPU multi-row fold) were #577, fixed in #587 and #604; M=2/3/4 are bit-identical to M=1 there now.
